### PR TITLE
feat: Perspective and Dynamic Sight

### PR DESCRIPTION
**Pull request by lgrossi**
## Summary

This pull request introduces a new component, `Perspective`, designed to represent the viewpoints or visible regions of entities within the game. It leverages raycasting and area targeting, based on the entity's position, to dynamically generate and manage visibility mechanics such as fogs, wards, team views, and more.

## Key Features

- **Perspective Component**: Represents an entity's viewpoint, encapsulating the regions it can "see."
- **ViewPoint Generation**: Utilizes `RadialViewPoint` to describe surrounding areas from which perspectives are derived.
- **Interception Calculation**: Processes and combines cached perspective data each frame to determine which positions are "in sight" for each entity.
- **Flexible Design**: Can be used not only for player view, but also for effects, spells, and others.

## Changes Overview

1. **Protobuf Update**: Renamed `blocks_visibility` flag to `blocks_sight` in `appearances.proto` to better reflect its purpose.
2. **TileFlags Module**: Introduced `tile_flags.rs` to handle tile-specific sight flags, enhancing the game's spatial awareness capabilities.
3. **Perspective Management**: Developed `perspective` module with structures and systems for managing entity viewpoints, including caching and sight processing mechanisms.
4. **Utility Updates**: Minor updates and optimizations to existing modules to support the new sight system.

## Implementation Details

- **Raycasting for Sight**: Implemented raycasting within view points to dynamically calculate "in sight" areas, considering obstructions like walls or terrain features.
- **Sight Intersection Caching**: Cached intersection calculations for efficient sight determination, reducing computational overhead.
- **Configurable ViewPoints**: `RadialViewPoint` allows for easy configuration of sight ranges and angles, offering flexibility in defining entity perspectives.
- **Sight Sharing**: Entities can share their viewpoints with others, simulating team sight and shared vision mechanics.

## Testing

Added comprehensive unit tests for the new functionality, focusing on the correct generation of perspectives and their expected behavior in various scenarios. Stress tests were also conducted to ensure performance stability under heavy loads.

[Screencast from 2024-04-01 22-22-58.webm](https://github.com/opentibiabr/Ryot/assets/34237492/750822e3-653c-42da-bc9a-281306633763)
[Screencast from 2024-04-01 01-40-06.webm](https://github.com/opentibiabr/Ryot/assets/34237492/2a23bcb3-f69e-41ff-ae0c-764c8e457cb6)

## Benchmark Results

Performance metrics for creating and checking perspectives, measured in perspectives processed per second (perspectives/s).

## Benchmark Results Summary

Performance metrics for creating and checking perspectives, measured in perspectives processed per second.

| Configuration                                | Creation (perspectives/s) | Checking (perspectives/s) |
|----------------------------------------------|---------------------------|---------------------------|
| Circular Perspectives (Radius 5 Tiles)       | 279,305                   | 405,037                   |
| Circular Perspectives (Radius 3 Tiles)       | 579,248                   | 1,090,088                 |
| Sector Perspectives (90 Degrees, Radius 5 Tiles) | 1,181,895             | 1,649,160                 |
| Sector Perspectives (90 Degrees, Radius 3 Tiles) | 2,202,938             | 3,437,943                 |

In our application's testing, we successfully rendered 3,000 to 5,000 circular perspectives, each 5x5 tiles, with live update depending on the tile content. We used gizmos as debug mode and it presented satisfactory performance,. This demonstrates our system's capability to manage complex visual environments smoothly, ensuring an engaging gameplay experience.

### CODE REVIEW

File: ryot/src/bevy_ryot/drawing/brushes/line.rs [hunk 1]
```diff
@@ -37,38 +37,9 @@ pub fn line<B: BrushItem>(params: BrushParams<B>, center: B) -> Vec<B> {
...
 
-    for pos in bresenhams_line(start_pos, center_pos) {
+    for pos in start_pos.bresenhams_line(center_pos) {
         elements.push(B::from_position(center, pos));
     }
 
     elements
 }
-
-fn bresenhams_line(a: TilePosition, b: TilePosition) -> Vec<TilePosition> {
```
Comment (luan) [inline ryot/src/bevy_ryot/drawing/brushes/line.rs:47]: Gesundheit

File: ryot/src/bevy_ryot/mod.rs [hunk 2]
```diff
@@ -51,14 +49,29 @@ pub mod drawing;
...
-use self::sprites::SpriteMaterial;
-
 pub static RYOT_ANCHOR: Anchor = Anchor::BottomRight;
 pub static GRID_LAYER: Layer = Layer::Hud(0);
 
+/// A generic cache structure leveraging `HashMap` for storing and quickly accessing data.
+/// This structure is particularly useful for caching expensive computations, assets, or
+/// other data for rapid retrieval.
+#[derive(Resource, Default, Deref, DerefMut)]
+pub struct Cache<K, V>(HashMap<K, V>);
```
Comment (luan) [inline ryot/src/bevy_ryot/mod.rs:64]: sometimes I wish you could just tell rust to make HashMap a `Resource` :|

File: ryot/src/bevy_ryot/perspective/tests/view_point_test.rs [hunk 3]
```diff
@@ -0,0 +1,312 @@
+use crate::bevy_ryot::perspective::*;
```
Comment (luan) [inline ryot/src/bevy_ryot/perspective/tests/view_point_test.rs:1]: did you consider using property testing here too? Similar to what we have in the positon module?

**Review by luan — APPROVED** [review]

File: ryot/src/bevy_ryot/perspective/tests/view_point_test.rs (hunk 3 above)
Comment (lgrossi) [reply ryot/src/bevy_ryot/perspective/tests/view_point_test.rs:1]: Added a few relevant ones, good call.

**Pull request merged**